google/utils.go

```diff
@@ -408,3 +365,252 @@ func extractFirstMapConfig(m []interface{}) map[string]interface{} {
+	natIP := ""
+	for i, ac := range accessConfigs {
+		flattened[i] = map[string]interface{}{
+			"nat_ip":          ac.NatIP,
+			"assigned_nat_ip": ac.NatIP,
+		}
+		if natIP == "" {
+			natIP = ac.NatIP
+		}
+	}
+	return flattened, natIP
+}
+
+func flattenNetworkInterfaces(networkInterfaces []*computeBeta.NetworkInterface) ([]map[string]interface{}, string, string, string) {
+	flattened := make([]map[string]interface{}, len(networkInterfaces))
+	var region, internalIP, externalIP string
+
+	for i, iface := range networkInterfaces {
+		var ac []map[string]interface{}
+		ac, externalIP = flattenAccessConfigs(iface.AccessConfigs)
+
+		var project string
+		project, region = getProjectAndRegionFromSubnetworkLink(iface.Subnetwork)
+
+		flattened[i] = map[string]interface{}{
+			"address":            iface.NetworkIP,
+			"network_ip":         iface.NetworkIP,
+			"network":            iface.Network,
+			"subnetwork":         iface.Subnetwork,
+			"subnetwork_project": project,
+			"access_config":      ac,
+			"alias_ip_range":     flattenAliasIpRange(iface.AliasIpRanges),
+		}
+		// Instance template interfaces never have names, so they're absent
+		// in the instance template network_interface schema. We want to use the
+		// same flattening code for both resource types, so we avoid trying to
+		// set the name field when it's not set at the GCE end.
+		if iface.Name != "" {
+			flattened[i]["name"] = iface.Name
+		}
+		if internalIP == "" {
+			internalIP = iface.NetworkIP
+		}
+	}
+	return flattened, region, internalIP, externalIP
+}
+
+func expandAccessConfigs(configs []interface{}) []*computeBeta.AccessConfig {
+	acs := make([]*computeBeta.AccessConfig, len(configs))
+	for i, raw := range configs {
+		data := raw.(map[string]interface{})
+		acs[i] = &computeBeta.AccessConfig{
+			Type:  "ONE_TO_ONE_NAT",
+			NatIP: data["nat_ip"].(string),
+		}
+	}
+	return acs
+}
+
+func expandNetworkInterfaces(d *schema.ResourceData, config *Config) ([]*computeBeta.NetworkInterface, error) {
+	project, err := getProject(d, config)
+	if err != nil {
+		return nil, err
+	}
+	region, err := getRegion(d, config)
+	if err != nil {
+		return nil, err
+	}
+
+	configs := d.Get("network_interface").([]interface{})
+	ifaces := make([]*computeBeta.NetworkInterface, len(configs))
+	for i, raw := range configs {
+		data := raw.(map[string]interface{})
+
+		network := data["network"].(string)
+		subnetwork := data["subnetwork"].(string)
+		if (network == "" && subnetwork == "") || (network != "" && subnetwork != "") {
+			return nil, fmt.Errorf("exactly one of network or subnetwork must be provided")
+		}
+
+		nf, err := ParseNetworkFieldValue(network, d, config)
+		if err != nil {
+			return nil, fmt.Errorf("cannot determine selflink for subnetwork '%s': %s", subnetwork, err)
+		}
+
+		subnetworkProject := data["subnetwork_project"].(string)
+		subnetLink, err := getSubnetworkLink(config, project, region, subnetworkProject, subnetwork)
```

Comment:
Were you intending on having this use ParseSubnetworkFieldValue? That got added in this PR but isn't used anywhere.

Reply:
https://github.com/terraform-providers/terraform-provider-google/pull/639#discussion_r151016103

> I added some code to parse regional fields, with the intention of using the same helpers for subnetworks. Unfortunately in the context of an instance/instance template the subnetwork's project isn't necessarily the same as the instance's project due to XPN, so I don't think the existing helpers will work.

I did originally, but then it was too much effort per the above comment. :) I think it might be worth doing, but it seems like it might be a good candidate for a separate PR given that `ParseSubnetworkFieldValue` would not work as currently implemented. I'm happy to drop `ParseSubnetworkFieldValue` and the regional field parsing code from this PR since it's unused if that's preferred.

Reply:
Ah ok. At this point my goal is to just get this PR in, so don't worry about it